src/IO/ForkWriteBuffer.cpp

```diff
@@ -0,0 +1,53 @@
+#include <IO/ForkWriteBuffer.h>
+#include <Common/Exception.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_CREATE_IO_BUFFER;
+}
+
+ForkWriteBuffer::ForkWriteBuffer(WriteBufferPtrs && sources_)
+    : WriteBuffer(nullptr, 0), sources(std::move(sources_))
+{
+    if (sources.empty())
+    {
+        throw Exception("Expected non-zero number of buffers for `ForkWriteBuffer`", ErrorCodes::CANNOT_CREATE_IO_BUFFER);
+    }
+    set(sources.front()->buffer().begin(), sources.front()->buffer().size());
+}
+
+
+void ForkWriteBuffer::nextImpl()
+{
+    sources.front()->position() = position();
+
+    try
+    {
+        auto & source_buffer = sources.front();
+        for (auto it = sources.begin() + 1; it != sources.end(); ++it)
+        {
+            auto & buffer = *it;
+            buffer->write(source_buffer->buffer().begin(), source_buffer->offset());
+            buffer->next();
+        }
+        source_buffer->next();
+    }
+    catch (Exception & exception)
+    {
+        exception.addMessage("While writing to ForkWriteBuffer");
+        throw;
+    }
+
+}
+
+ForkWriteBuffer::~ForkWriteBuffer()
+{
+    finalize();
+    set(nullptr, 0);
```

Comment:
I think this `set` is not needed

Reply:
From my understanding, after `~ForkWriteBuffer()`, the `~()WriteBuffer()-> finalize()->next()` will be called. In which `offset() `is checked and if it's not zero then `nextImp()` of WriteBuffer is called. So I added this to clear the Buffer so that we dont call nextImpl(). If we were to remove the finalize() call from above comment, then we would need to clear buffer / set `pos = working_buffer.begin()`. (currently `next() ` call always clears it)
All of this is assuming we are in ForkWriteBuffer destructor with some data in buffer yet to be written. Please let me know your thoughts

Reply:
> From my understanding, after ~ForkWriteBuffer(), the ~()WriteBuffer()-> finalize()->next() will be called.

No, because the destructor of `WriteBuffer` is defined as `virtual ~WriteBuffer() = default;`.

Reply:
Oh thats a good point. Silly of me to have missed it. 
I have addressed it.